tensorflow/java/src/gen/cc/op_parser.cc

```diff
@@ -0,0 +1,417 @@
+  if (visited_attrs_.find(attr_def.name()) != visited_attrs_.cend()) {
+    return;
+  }
+  bool iterable = false;
+  Type data_type = DataTypeOf(attr_def, &iterable);
+  // generic attributes should be passed as an explicit type
+  bool explicit_type = data_type.kind() == Type::GENERIC && !iterable;
+  Type type = explicit_type ?
+      Type::Class("Class").add_parameter(data_type) : data_type;
+  if (iterable) {
+    type = Type::ListOf(data_type);
+  }
+  OpSpec::Operand attr(attr_api.name(),
+      Variable::Create(SnakeToCamelCase(attr_api.rename_to()), type),
+      data_type,
+      ParseDocumentation(attr_api.description()),
+      iterable);
+  // attributes with a default value are optional
+  if (attr_api.has_default_value() && !explicit_type) {
+    op->add_option(attr);
+  } else {
+    op->add_attribute(attr);
+  }
+  visited_attrs_.insert(std::make_pair(attr_api.name(), data_type));
+}
+
+Type OpParser::DataTypeOf(const OpDef_ArgDef& arg, bool* iterable_out) {
+  if (!arg.number_attr().empty()) {
+    visited_attrs_.insert(std::make_pair(arg.number_attr(), Type::Int()));
+    *iterable_out = true;
+  }
+  if (arg.type() != DataType::DT_INVALID) {
+    // resolve type from DataType
+    switch (arg.type()) {
+      case DataType::DT_BOOL:
+        return Type::Class("Boolean");
+
+      case DataType::DT_STRING:
+        return Type::Class("String");
+
+      case DataType::DT_FLOAT:
+        return Type::Class("Float");
+
+      case DataType::DT_DOUBLE:
+        return Type::Class("Double");
+
+      case DataType::DT_UINT8:
+        return Type::Class("UInt8", "org.tensorflow.types");
+
+      case DataType::DT_INT32:
+        return Type::Class("Integer");
+
+      case DataType::DT_INT64:
+        return Type::Class("Long");
+
+      case DataType::DT_RESOURCE:
+        // TODO(karllessard) create a Resource utility class that could be
+        // used to store a resource and its type (passed in a second argument).
+        // For now, we need to force a wildcard and we will unfortunately lose
+        // track of the resource type.
+        return Type::Wildcard();
+
+      default:
+        break;
+    }
+  } else {
+    // resolve type from type attribute
+    string attr_name = arg.type_attr();
+    if (attr_name.empty()) {
+      attr_name = arg.type_list_attr();
+      if (!attr_name.empty()) {
+        *iterable_out = true;
+        Type type = Type::Wildcard();
+        visited_attrs_.insert(std::make_pair(attr_name, type));
+        return type;
+      }
+    }
+    for (const auto& attr : op_def_.attr()) {
+      if (attr.name() == attr_name) {
+        Type type = DataTypeOf(attr, iterable_out);
+        visited_attrs_.insert(std::make_pair(attr_name, type));
+        return type;
+      }
+    }
+  }
+  LOG(WARNING) << "Data type for arg \"" << arg.name() << "\" is unknown";
+  return Type::Wildcard();
+}
+
+Type OpParser::DataTypeOf(const OpDef_AttrDef& attr, bool* iterable_out) {
```

Comment:
I tried to avoir `TypeOf` because the type of an input, for example, can be `List<Operand<String>>` while the value return here is only the tensor's one (`String` in this case). Maybe `TensorTypeOf` then?

Reply:
I see. I think `TypeOf` with a comment is fine. Or maybe `TypeOfElement`?